agent/agent.go

```diff
@@ -0,0 +1,635 @@
+
+			fileType = request.GetType()
+			fileHandle, err = o.initFile(fileType, request.GetFilename(), request.GetOverwrite())
+			if err != nil {
+				return err
+			}
+			lastChunkIdx = request.GetChunkIdx() - 1
+		}
+
+		chunkIdx := request.GetChunkIdx()
+		if chunkIdx != int32(1+lastChunkIdx) {
+			return fmt.Errorf("received chunkIdx: %d after %d", chunkIdx, lastChunkIdx)
+		}
+		lastChunkIdx = chunkIdx
+
+		numChunks++
+		bytes := request.GetChunkBytes()
+		checksum = crc32.Update(checksum, crc32.IEEETable, bytes)
+
+		numWritten, err := fileHandle.Write(bytes)
+		if err != nil {
+			return err
+		}
+
+		if numWritten != len(bytes) {
+			return fmt.Errorf("unable to write bytes, expected: %d, observed: %d", len(bytes), numWritten)
+		}
+
+		if streamErr == io.EOF {
+			o.markFileDone(fileType, fileHandle.Name())
+			return stream.SendAndClose(&m3em.TransferResponse{
+				FileChecksum:   checksum,
+				NumChunksRecvd: int32(numChunks),
+			})
+		}
+	}
+}
+
+type opAgentHeartBeater struct {
+	sync.RWMutex
+	iopts             instrument.Options
+	agent             *opAgent
+	running           bool
+	wg                sync.WaitGroup
+	msgChan           chan heartbeatMsg
+	conn              *grpc.ClientConn
+	client            hb.HeartbeaterClient
+	operatorUUID      string
+	heartbeatEndpoint string
+}
+
+func newHeartbeater(
+	agent *opAgent,
+	opUUID string,
+	hbEndpoint string,
+	iopts instrument.Options,
+) (*opAgentHeartBeater, error) {
+	conn, err := grpc.Dial(hbEndpoint, grpc.WithTimeout(defaultHeartbeatConnTimeout), grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+	client := hb.NewHeartbeaterClient(conn)
+	return &opAgentHeartBeater{
+		iopts:             iopts,
+		agent:             agent,
+		msgChan:           make(chan heartbeatMsg),
+		conn:              conn,
+		client:            client,
+		operatorUUID:      opUUID,
+		heartbeatEndpoint: hbEndpoint,
+	}, nil
+}
+
+func (h *opAgentHeartBeater) isRunning() bool {
+	h.RLock()
+	running := h.running
+	h.RUnlock()
+	return running
+}
+
+func (h *opAgentHeartBeater) defaultHeartbeat() hb.HeartbeatRequest {
+	h.RLock()
+	req := hb.HeartbeatRequest{
+		OperatorUuid:   h.operatorUUID,
+		ProcessRunning: h.agent.Running(),
+	}
+	h.RUnlock()
+	return req
+}
+
+func (h *opAgentHeartBeater) start(d time.Duration) {
+	h.Lock()
+	h.running = true
+	h.Unlock()
+	h.wg.Add(1)
+	go h.heartbeatLoop(d)
+}
+
+func (h *opAgentHeartBeater) sendHealthyHeartbeat() {
+	beat := h.defaultHeartbeat()
+	beat.Code = hb.HeartbeatCode_HEALTHY
+	h.sendHeartbeat(&beat)
+}
+
+func (h *opAgentHeartBeater) heartbeatLoop(d time.Duration) {
+	defer h.wg.Done()
+	defer func() {
+		h.Lock()
+		h.running = false
+		h.Unlock()
+	}()
+
+	// explicitly send first heartbeat as soon as we start
+	h.sendHealthyHeartbeat()
+
+	var (
+		timer     = time.NewTimer(d)
+		heartbeat = true
+	)
+	defer timer.Stop()
+
+	for heartbeat {
+		select {
+		case msg := <-h.msgChan:
+			if msg.stop {
+				heartbeat = false
```

Comment:
can't, `break` would apply to the innermost select statement here. [ref](https://golang.org/ref/spec#Break_statements)